server/src/main/java/org/apache/druid/metadata/IndexerSQLMetadataStorageCoordinator.java

```diff
@@ -202,16 +197,30 @@ public List<Pair<DataSegment, String>> retrieveUsedSegmentsAndCreatedDates(Strin
               .createQuery(queryString)
               .bind("dataSource", dataSource);
 
-          SqlSegmentsMetadataQuery.bindQueryIntervals(query, intervals);
+          if (compareAsString) {
+            SqlSegmentsMetadataQuery.bindQueryIntervals(query, intervals);
+          }
 
-          return query
-              .map((int index, ResultSet r, StatementContext ctx) ->
-                       new Pair<>(
-                           JacksonUtils.readValue(jsonMapper, r.getBytes("payload"), DataSegment.class),
-                           r.getString("created_date")
-                       )
-              )
-              .list();
+          return query.map((int index, ResultSet r, StatementContext ctx) ->
+                               new Pair<>(
+                                   JacksonUtils.readValue(jsonMapper, r.getBytes("payload"), DataSegment.class),
+                                   r.getString("created_date")
+                               )
+                      )
+                      .list()
```

Comment:
I think I would prefer to build this list and then filter it later in a separate statement only if `intervals` is non-empty and we couldn't compare as strings.

So something like:

```java
List<Pair<>> list = query.map().list();

if (intervals.isEmpty() || compareAsString) {
  return list;
} else {
  return list.stream().filter().collect();
}
```

Reply:
Done